packages/builder/src/pages/builder/app/[application]/design/_components/NewScreen/CreateScreenModal.svelte

```diff
@@ -44,136 +43,83 @@
-    const screens = get(screenStore).screens.filter(
-      s => s.routing.roleId === screenAccessRole
-    )
-    return !!screens.find(s => s.routing?.route === url)
-  }
-
-  // Constructs a candidate URL for a new screen, appending a given suffix to the
-  // screen's URL
-  // e.g. "/sales/:id" => "/sales-1/:id"
-  const makeCandidateUrl = (screen, suffix) => {
-    let url = screen.routing?.route || ""
-    if (url.startsWith("/")) {
-      url = url.slice(1)
-    }
-    if (!url.includes("/")) {
-      return `/${url}-${suffix}`
-    } else {
-      const split = url.split("/")
-      return `/${split[0]}-${suffix}/${split.slice(1).join("/")}`
-    }
   }
 
   const onSelectDatasources = async () => {
     if (mode === "form") {
       formTypeModal.show()
-    } else if (mode === "grid") {
-      await createGridScreen()
-    } else if (mode === "gridDetails") {
-      await createGridDetailsScreen()
+    } else if (mode === "table") {
+      tableTypeModal.show()
     }
   }
 
-  const createBlankScreen = async ({ screenUrl }) => {
-    const screenTemplate = blankScreen(screenUrl)
-    const screen = await createScreen(screenTemplate)
-    await addNavigationLink(screenTemplate)
+  const createBlankScreen = async ({ route }) => {
+    const screenTemplates = screenTemplating.blank({ route, screens })
 
-    loadNewScreen(screen)
+    const newScreens = await createScreens(screenTemplates)
+    loadNewScreen(newScreens[0])
   }
 
-  const createGridScreen = async () => {
-    let firstScreen = null
-
-    for (let tableOrView of selectedTablesAndViews) {
-      const screenTemplate = gridScreen(
+  const createTableScreen = async type => {
+    const screenTemplates = selectedTablesAndViews.flatMap(tableOrView =>
+      screenTemplating.table({
+        screens,
         tableOrView,
-        permissions[tableOrView.id]
-      )
-
-      const screen = await createScreen(screenTemplate)
-      await addNavigationLink(screen)
-
-      firstScreen ??= screen
-    }
-
-    loadNewScreen(firstScreen)
-  }
-
-  const createGridDetailsScreen = async () => {
-    let firstScreen = null
-
-    for (let tableOrView of selectedTablesAndViews) {
-      const screenTemplate = gridDetailsScreen(
-        tableOrView,
-        permissions[tableOrView.id]
-      )
-
-      const screen = await createScreen(screenTemplate)
-      await addNavigationLink(screen)
-
-      firstScreen ??= screen
-    }
+        type,
+        permissions: permissions[tableOrView.id],
+      })
+    )
 
-    loadNewScreen(firstScreen)
+    const newScreens = await createScreens(screenTemplates)
+    loadNewScreen(newScreens[0])
   }
 
-  const createFormScreen = async formType => {
-    let firstScreen = null
-
-    for (let tableOrView of selectedTablesAndViews) {
-      const screenTemplate = formScreen(
+  const createFormScreen = async type => {
+    const screenTemplates = selectedTablesAndViews.flatMap(tableOrView =>
```

Comment:
Sorry, I'd taken note to mention this in the main body of the PR, but looks like I forgot.

It's just so that all these generator functions have the same interface, so because one of them needs to return an array of screens, it makes sense (to me, at least) just to have them all follow that pattern. I think having them all use/return the same interface will save us some pain once we introduced more templates and we're forced to DRY up the new screen modal flow.